### PR TITLE
added new features

### PR DESCRIPTION
**Pull request by jstsanam**

### CODE REVIEW

File: controllers/listingController.js [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+        req.flash("error", "Listing doesn't exist anymore 🤷🏻‍♀️");
+        res.redirect("/listings");
+    }
+    res.render("listings/show.ejs", { listing });
+}
+
+const createNewListing = async (req, res, next) => {
+    let url = req.file.path;
+    let filename = req.file.filename;
+    console.log(url, "-------", filename);
```
Comment (sanchit203) [inline controllers/listingController.js]: do we need this console log?

File: controllers/listingController.js [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    const newListing = new Listing(req.body.listing);
+    newListing.owner = req.user._id;
+    newListing.image = { url, filename };
+    await newListing.save();
+    req.flash("success", "New listing added successfully ✅");
```
Comment (sanchit203) [inline controllers/listingController.js:37]: for now these are okay but if you want to write more better code, you can create enums for status... and constants for FlashMessages and use that everywhere.
So basicallay you will have a constants folders which will contain a file FlashMessages.js that file will look something like...

`export const NEW_LISTING_SUCCESS_MESSAGE = "New listing added successfully ✅"` and all other messages as well.

For enum you can create a enum folder and create a enum file there

File: controllers/listingController.js [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+    res.redirect("/listings");
```
Comment (sanchit203) [inline controllers/listingController.js:38]: In constants folder you can have a file called `EjsRoutes` which will contain all the constants for EjsRoutes like this listings

`export const LISTING_EJS_TEMPLATE = "/listings"`

File: controllers/listingController.js [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+}
+
+const showEditListing = async (req, res) => {
+    let { id } = req.params;
+    const listing = await Listing.findById(id);
+    if (!listing) {
+        req.flash("error", "Listing doesn't exist anymore 🤷🏻‍♀️");
```
Comment (sanchit203) [inline controllers/listingController.js:45]: similary you can move these to enums and messages

File: controllers/listingController.js [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+        res.redirect("/listings");
+    }
+    res.render("listings/edit.ejs", { listing });
```
Comment (sanchit203) [inline controllers/listingController.js:48]: same these in EjsTemplateRoutes

File: controllers/listingController.js [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+}
+
+const editListing = async (req, res) => {
+    let { id } = req.params;
+    await Listing.findByIdAndUpdate(id, { ...req.body.listing });
+    req.flash("success", "Listing updated sucessfully ✅");
+    res.redirect(`/listings/${id}`);
```
Comment (sanchit203) [inline controllers/listingController.js:55]: For this you can look into how to move this in a constant and create a route while using the const. Do a bit of google search

File: models/listingSchema.js [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+            ref: "Review",
+        }
+    ],
+    owner: {
+        type: Schema.Types.ObjectId,
+        ref: "User"
+    }
+});
+
+//will delete reviews associated with listings
```
Comment (sanchit203) [inline models/listingSchema.js]: missing <space> after the // and before will.. also this is not a relevant comment.. relevant comment would be
`// Deletes the revies associated with listings` no need of will.. Do the same thing for every comment

File: routes/indexRoute.js [hunk 8]
```diff
@@ -1,9 +1,8 @@
 import express from "express";
+import { index } from "../controllers/indexController.js";
 
 const indexRoute = express.Router();
 
-indexRoute.get("/", (req, res) => {
-    res.render("./listings/welcome.ejs");
-});
+indexRoute.get("/", index);
```
Comment (sanchit203) [inline routes/indexRoute.js:6]: you can create one more constant file call Routes.js and have all the routes exported as a constant in there. One file for Routes that is hit from the browser and one is for the EjsTemplateRoutes

File: routes/listingRoute.js [hunk 9]
```diff
@@ -1,113 +1,74 @@
...
+  editListing,
+  deleteListing,
+  createReview,
+  deleteReview
+} from "../controllers/listingController.js";
+import multer from "multer";
+import { storage } from "../cloudConfig.js";
 
 const listingRoute = express.Router();
+//multer is node.js middleware for handling multipart/form-data (specially upload media)
```
Comment (sanchit203) [inline routes/listingRoute.js]: space!!

File: routes/listingRoute.js [hunk 10]
```diff
@@ -1,113 +1,74 @@
...
+const upload = multer({ storage });
 
-listingRoute.get("/",
-  wrapAsync(async (req, res) => {
-    const allListings = await Listing.find({});
-    res.render("listings/index.ejs", { allListings });
-  }));
+listingRoute
+  .route("/")
```
Comment (sanchit203) [inline routes/listingRoute.js:26]: can use the indexRoute exported from Routes.js file of constant

File: routes/listingRoute.js [hunk 11]
```diff
@@ -1,113 +1,74 @@
...
+  .get(wrapAsync(listingIndexPage))
+  .post(
+    isAuthenticated,
+    upload.single('listing[image]'),
+    validateListing,
+    wrapAsync(createNewListing)
+  );
 
 listingRoute.get("/new",
```
Comment (sanchit203) [inline routes/listingRoute.js:35]: can create NEW_LISTING_ROUTE in Route.js

**Review by sanchit203 — CHANGES_REQUESTED** [review]

File: controllers/listingController.js (hunk 1 above)
Comment (jstsanam) [reply controllers/listingController.js]: no

File: controllers/listingController.js (hunk 2 above)
Comment (jstsanam) [reply controllers/listingController.js:37]: will do ✅

File: controllers/listingController.js (hunk 3 above)
Comment (jstsanam) [reply controllers/listingController.js:38]: will do ✅

File: controllers/listingController.js (hunk 4 above)
Comment (jstsanam) [reply controllers/listingController.js:45]: will do ✅

File: controllers/listingController.js (hunk 5 above)
Comment (jstsanam) [reply controllers/listingController.js:48]: will do ✅

File: controllers/listingController.js (hunk 6 above)
Comment (jstsanam) [reply controllers/listingController.js:55]: will do ✅

File: models/listingSchema.js (hunk 7 above)
Comment (jstsanam) [reply models/listingSchema.js]: done

File: routes/indexRoute.js (hunk 8 above)
Comment (jstsanam) [reply routes/indexRoute.js:6]: will do ✅

File: routes/listingRoute.js (hunk 9 above)
Comment (jstsanam) [reply routes/listingRoute.js]: done

File: routes/listingRoute.js (hunk 10 above)
Comment (jstsanam) [reply routes/listingRoute.js:26]: will do ✅

File: routes/listingRoute.js (hunk 11 above)
Comment (jstsanam) [reply routes/listingRoute.js:35]: will do ✅

**Comment by jstsanam** [comment]
Will do these changes later, you can merge the files for now.

**Review by sanchit203 — APPROVED** [review]

**Pull request merged**